server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -46,42 +75,112 @@ public void beforeRefresh() throws IOException {
 
     /**
      * Upload new segment files created as part of the last refresh to the remote segment store.
-     * The method also deletes segment files from remote store which are not part of local filesystem.
+     * This method also uploads remote_segments_metadata file which contains metadata of each segment file uploaded.
      * @param didRefresh true if the refresh opened a new reference
-     * @throws IOException in case of I/O error in reading list of local files
      */
     @Override
-    public void afterRefresh(boolean didRefresh) throws IOException {
-        if (didRefresh) {
-            Set<String> localFiles = Set.of(storeDirectory.listAll());
-            localFiles.stream().filter(file -> !filesUploadedToRemoteStore.contains(file)).forEach(file -> {
-                try {
-                    remoteDirectory.copyFrom(storeDirectory, file, file, IOContext.DEFAULT);
-                    filesUploadedToRemoteStore.add(file);
-                } catch (NoSuchFileException e) {
-                    logger.info(
-                        () -> new ParameterizedMessage("The file {} does not exist anymore. It can happen in case of temp files", file),
-                        e
-                    );
-                } catch (IOException e) {
-                    // ToDO: Handle transient and permanent un-availability of the remote store (GitHub #3397)
-                    logger.warn(() -> new ParameterizedMessage("Exception while uploading file {} to the remote segment store", file), e);
-                }
-            });
+    public void afterRefresh(boolean didRefresh) {
+        synchronized (this) {
+            try {
+                if (indexShard.shardRouting.primary()) {
```

Comment:
What happens if index shard was demoted to replica from primary? Shouldn't `RemoteStoreRefreshListener` reflect that? Otherewise we are risking to lose track of shard state when it is flipping between primary <-> replica

Reply:
Only primary should be uploading the segments. If replica is demoted, this check makes sure that we stop uploading segments from replica. Ideally, `NRTReplicationEngine` should take of this. This `if` condition is added as a safety check.

Reply:
So in this case the usage of `isPrimary` property is not useful since it never reflects the actual state of the shard (if demoted from primary). I think we could drop it than

Reply:
If the shard is demoted and promoted again, we need to make sure that we call `RemoteSegmentStoreDirectory.init()` to get the latest list of files from remote store (because it could be changed by another primary in the meantime).

Reply:
Ok, I will try to go step by step by step:
 - `indexShard.shardRouting.primary() == true`: if `isPrimary` was `false`, we set it to `true` and proceed with `init()`
 - the shard flips from primary to replica, `indexShard.shardRouting.primary() == false`, nothing happens to `isPrimary`, it is still, `true` **(which is wrong)**
 - the shard flips from  replica to primary, `indexShard.shardRouting.primary() == true` and `isPrimary` is also `true`  - the initialization call to `init()` won't happen anymore because the flip to replica was not reflected

Does it make sense?

Reply:
Yeah, this is definitely a bug. Thanks for catching it. I will fix it and will add tests around it as you suggested in another comment.

Reply:
@reta I changed the implementation and added tests. Please take a look again.

Reply:
Looks good, thanks!